# Analysis on HACKER NEWS Site
In this project, we'll work with a data set of submissions to popular technology site Hacker News.

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set [here](https://www.kaggle.com/hacker-news/hacker-news-posts), Below are descriptions of the columns:

* id: The unique identifier from Hacker News for the post
* title: The title of the post
* url: The URL that the posts links to, if it the post has a URL
* num_points: The number of points the post acquired, calculated as the total * * number of upvotes minus the total number of downvotes
* num_comments: The number of comments that were made on the post
* author: The username of the person who submitted the post
* created_at: The date and time at which the post was submitted

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

Ask HN: How to improve my personal website?                                         
Ask HN: Am I the only one outraged by Twitter shutting down share counts?                   
Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'       
Show HN: Something pointless I made                                                 
Show HN: Shanhu.io, a programming playground powered by e8vm                               

We'll compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

In [2]:
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


Notice that the first list in the inner lists contains the column headers, and the lists after contain the data for one row. In order to analyze our data, we need to first remove the row containing the column headers. Let's remove that first row next.

In [3]:
headers = hn[0]
hn = hn[1:]
print("The column header: ")
print(headers)
print("The values of first five rows: ")
print(hn[:5])

The column header: 
['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
The values of first five rows: 
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2

## Extracting Ask HN and Show HN Posts

we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles. So we are going to extrack the data from hn list.


In [4]:
ask_posts =[]
show_posts =[]
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
print(ask_posts[:4])

    

1744
1162
17194
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20']]


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

In [5]:
total_ask_comments = 0

for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments = total_ask_comments + num_comments
   
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print("ask comments average: ",avg_ask_comments)

total_show_comments  = 0
row = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments = total_show_comments + num_comments
    
    
avg_show_comments = total_show_comments / len(show_posts)
print("show comments average: ",avg_show_comments)
    

ask comments average:  14.038417431192661
show comments average:  10.31669535283993


We can determined that, on average, ask posts receive more comments than show posts. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts. 

## Finding the Amount of Ask Posts and Comments by Hour Created

 we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

* Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
* Calculate the average number of comments ask posts receive by hour created.

In [6]:
import datetime as dt

result_list = []

for row in ask_posts:
    created_at = row[6]
    no_of_comments = int(row[4])
    result_list.append([created_at,no_of_comments])

#print(result_list)
counts_by_hour ={}
comments_by_hour = {}

for row in result_list:
    date_object = dt.datetime.strptime(row[0],"%m/%d/%Y %H:%M")
    hour_extract = dt.datetime.strftime(date_object,"%H")
    #print(hour_extract)
    if hour_extract not in counts_by_hour:
        counts_by_hour[hour_extract] =1
        comments_by_hour[hour_extract] = row[1]
    else:
        counts_by_hour[hour_extract] += 1
        comments_by_hour[hour_extract] += row[1]


In [7]:
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

avg_by_hour

[['07', 7.852941176470588],
 ['15', 38.5948275862069],
 ['20', 21.525],
 ['18', 13.20183486238532],
 ['00', 8.127272727272727],
 ['08', 10.25],
 ['11', 11.051724137931034],
 ['02', 23.810344827586206],
 ['23', 7.985294117647059],
 ['22', 6.746478873239437],
 ['06', 9.022727272727273],
 ['13', 14.741176470588234],
 ['16', 16.796296296296298],
 ['12', 9.41095890410959],
 ['09', 5.5777777777777775],
 ['10', 13.440677966101696],
 ['21', 16.009174311926607],
 ['03', 7.796296296296297],
 ['14', 13.233644859813085],
 ['01', 11.383333333333333],
 ['19', 10.8],
 ['05', 10.08695652173913],
 ['17', 11.46],
 ['04', 7.170212765957447]]

## Sorting and Printing Values from a List of Lists



In [11]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap

[[7.852941176470588, '07'], [38.5948275862069, '15'], [21.525, '20'], [13.20183486238532, '18'], [8.127272727272727, '00'], [10.25, '08'], [11.051724137931034, '11'], [23.810344827586206, '02'], [7.985294117647059, '23'], [6.746478873239437, '22'], [9.022727272727273, '06'], [14.741176470588234, '13'], [16.796296296296298, '16'], [9.41095890410959, '12'], [5.5777777777777775, '09'], [13.440677966101696, '10'], [16.009174311926607, '21'], [7.796296296296297, '03'], [13.233644859813085, '14'], [11.383333333333333, '01'], [10.8, '19'], [10.08695652173913, '05'], [11.46, '17'], [7.170212765957447, '04']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [13]:
#sort values and find the top 5 ask post
print("\n")
print("Top 5 Hours for Ask Post Comments")
print("----------------------------------")

for row in sorted_swap[:5]:
    time = dt.time(hour=int(row[1])).strftime("%H:%M")
    print("{time}:{avg:.2f} average comments per post".format(time=time, avg=row[0]))



Top 5 Hours for Ask Post Comments
----------------------------------
15:00:38.59 average comments per post
02:00:23.81 average comments per post
20:00:21.52 average comments per post
16:00:16.80 average comments per post
21:00:16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

## Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.